src/instructlab/model/backends/backends.py

```diff
@@ -13,19 +56,253 @@ def __init__(
         self,
         logger: logging.Logger,
         model_path: pathlib.Path,
+        api_base: str,
         host: str,
         port: int,
+        process: multiprocessing.Process = None,
         **kwargs,
     ) -> None:
         self.logger = logger
         self.model_path = model_path
+        self.api_base = api_base
         self.host = host
         self.port = port
+        self.process = process
 
     @abc.abstractmethod
     def run(self):
-        """Run serving backend"""
+        """Run serving backend in foreground (ilab model serve)"""
+
+    def run_detached(self):
+        """Run serving backend in background ('ilab model chat' when server is not running)"""
 
     @abc.abstractmethod
     def shutdown(self):
         """Shutdown serving backend"""
+
+
+def is_model_gguf(model_path: pathlib.Path) -> bool:
+    """
+    Check if the file is a GGUF file.
+    Args:
+        model_path (Path): The path to the file.
+    Returns:
+        bool: True if the file is a GGUF file, False otherwise.
+    """
+    # Standard
+    import mmap
+    import struct
+
+    # Third Party
+    from gguf.constants import GGUF_MAGIC
+
+    with open(model_path, "rb") as f:
+        # Memory-map the file on the first 4 bytes (this is where the magic number is)
+        mmapped_file = mmap.mmap(f.fileno(), length=4, access=mmap.ACCESS_READ)
+
+        # Read the first 4 bytes
+        first_four_bytes = mmapped_file.read(4)
+
+        # Convert the first four bytes to an integer
+        first_four_bytes_int = struct.unpack("<I", first_four_bytes)[0]
+
+        # Close the memory-mapped file
+        mmapped_file.close()
+
+        return first_four_bytes_int == GGUF_MAGIC
+
+
+def validate_backend(backend: str) -> None:
+    """
+    Validate the backend.
+    Args:
+        backend (str): The backend to validate.
+    Raises:
+        ValueError: If the backend is not supported.
+    """
+    # lowercase backend for comparison in case of user input like 'Llama'
+    if backend.lower() not in SUPPORTED_BACKENDS:
+        raise ValueError(
+            f"Backend '{backend}' is not supported. Supported: {', '.join(SUPPORTED_BACKENDS)}"
+        )
+
+
+def determine_backend(model_path: pathlib.Path) -> str:
+    """
+    Determine the backend to use based on the model file properties.
+    Args:
+        model_path (Path): The path to the model file.
+    Returns:
+        str: The backend to use.
+    """
+    # Check if the model is a GGUF file
+
+    # If the model is a directory, it's a VLLM model - it's kinda weak, but it's a start
+    if model_path.is_dir() and sys.platform == "linux":
```

Comment:
As a follow up PR, we could auto detect `.safetensors` files, and the presence of certain specific `.json` files existing. I think a follow up issue for this could get created as soon as the PR gets merged, and we don't hold up on it.

Reply:
Sounds good as a follow-up item to me. Thanks!

Reply:
https://github.com/instructlab/instructlab/issues/1457